src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -102,15 +118,114 @@
+  }
+
+  /**
+   * Handles transition of given invoice to PAID status.
+   *
+   * @param invoice Invoice to paid
+   * @return CompletableFuture that indicates when transition is completed
+   */
+  private CompletableFuture<Void> payInvoice(Invoice invoice) {
+    return fetchInvoiceLinesByInvoiceId(invoice.getId())
+      .thenApply(this::groupInvoiceLinesByPoLineId)
+      .thenCompose(this::fetchPoLines)
+      .thenApply(this::updatePoLinesPaymentStatus)
+      .thenCompose(this::updateCompositePoLines)
+      .thenCompose(aVoid -> updateInvoiceRecord(invoice));
+  }
+
+
+  private CompletableFuture<List<InvoiceLine>> fetchInvoiceLinesByInvoiceId(String invoiceId) {
+    String query = "invoiceId==" + invoiceId;
+    // Assuming that the invoice will never contain more than Integer.MAX_VALUE invoiceLines.
+    return invoiceLineHelper.getInvoiceLines(Integer.MAX_VALUE, 0, query)
+      .thenApply(InvoiceLineCollection::getInvoiceLines);
+  }
+
+  private Map<String, List<InvoiceLine>>  groupInvoiceLinesByPoLineId(List<InvoiceLine> invoiceLines) {
+    if (CollectionUtils.isEmpty(invoiceLines)) {
+      return Collections.emptyMap();
+    }
+    return invoiceLines
+      .stream()
+      .collect(Collectors.groupingBy(InvoiceLine::getPoLineId));
+  }
+
+  private CompletableFuture<Map<CompositePoLine, List<InvoiceLine>>> fetchPoLines(Map<String, List<InvoiceLine>> poLineIdsWithInvoiceLines) {
+    List<CompletableFuture<CompositePoLine>> futures = poLineIdsWithInvoiceLines.keySet()
+      .stream()
+      .map(this::getPoLineById)
+      .collect(toList());
+
+    return VertxCompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+      .thenApply(v -> futures.stream()
+        .map(CompletableFuture::join)
+        .collect(Collectors.toMap(poLine -> poLine, poLine -> poLineIdsWithInvoiceLines.get(poLine.getId()))));
   }
+
+  private CompletableFuture<CompositePoLine> getPoLineById(String poLineId) {
+    return handleGetRequest(String.format(PO_LINE_BY_ID_ENDPOINT, poLineId, lang), httpClient, ctx, okapiHeaders, logger)
+      .thenApply(jsonObject -> jsonObject.mapTo(CompositePoLine.class))
+      .exceptionally(throwable -> {
+        logger.error("Exception retrieving poLine with id={}:", throwable, poLineId);
+        throw new HttpException(500, String.format("PoLine with id=%s not found", poLineId));
+      });
+  }
+
+  private List<CompositePoLine> updatePoLinesPaymentStatus(Map<CompositePoLine, List<InvoiceLine>> compositePoLinesWithInvoiceLines) {
+    List<CompositePoLine> compositePoLines =  new ArrayList<>(compositePoLinesWithInvoiceLines.keySet());
+     compositePoLines
+      .forEach(compositePoLine -> {
+        List<InvoiceLine> invoiceLines =  compositePoLinesWithInvoiceLines.get(compositePoLine);
+        compositePoLine.setPaymentStatus(getPoLinePaymentStatus(invoiceLines));
+      });
+    return compositePoLines;
```

Comment:
I think it makes sense to return here only list of PO Lines, which status should be changed

Reply:
Done